piff/knn_interp.py

```diff
@@ -0,0 +1,373 @@
+            # if no y, then interpolate
+            y = numpy.array([self.knn[key].predict(X) for key in self.attr_target]).T
+        if logger:
+            logger.debug('Regression shape: %s', y.shape)
+        # add misalignment shape (n_targets, 3)
+        # X is (n_samples, 2)
+        # y is (n_samples, n_targets)
+        y = y + self.misalignment[numpy.newaxis, :, 0] \
+                + X[:, 1, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 1] \
+                + X[:, 0, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 2]
+
+        return y
+
+    def writeSolution(self, fits, extname):
+        """Write the solution to a FITS binary table.
+
+        Save the knn params and the X and y arrays
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+
+        dtypes = [('X', self.X.dtype, self.X.shape), ('Y', self.y.dtype, self.y.shape),
+                  ('XPIXEL', self.Xpixel.dtype, self.Xpixel.shape),
+                  ('ATTR_TARGET', self.attr_target.dtype, self.attr_target.shape),
+                  ('ATTR_INTERP', self.attr_interp.dtype, self.attr_interp.shape),
+                  ('MISALIGNMENT', self.misalignment.dtype, self.misalignment.shape)]
+        data = numpy.empty(1, dtype=dtypes)
+        # assign
+        data['X'] = self.X
+        data['Y'] = self.y
+        data['XPIXEL'] = self.Xpixel
+        data['ATTR_TARGET'] = self.attr_target
+        data['ATTR_INTERP'] = self.attr_interp
+        data['MISALIGNMENT'] = self.misalignment
+
+        # put the knn params in the header?
+        header = self.kwargs
+
+        # write to fits
+        fits.write_table(data, extname=extname, header=header)
+
+    def readSolution(self, fits, extname):
+        """Read the solution from a FITS binary table.
+
+        The extension should contain the same values as are saved
+        in the writeSolution method.
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+        # header = fits[extname].read_header()
+        data = fits[extname].read()
+
+        # kwargs come from base class read()
+        # self.kwargs = header
+
+        # attr_target and attr_interp assigned in build
+        self.build(data['ATTR_INTERP'][0], data['ATTR_TARGET'][0])
+        # self.X and self.y assigned in _fit
+        self._fit(data['X'][0], data['Y'][0])
+        self.misalign_wavefront(data['MISALIGNMENT'][0])
+
+        # other attributes
+        self.Xpixel = data['XPIXEL'][0]
+
+
+# pretty sure the correct way to do this is in the WCS framework, but let's
+# move forward first:
+def pixel_to_focal(stardata):
+    """Take stardata and add focal plane position to properties
+
+    :param stardata:    The stardata with property 'ccdnum'
+
+    :returns stardata:  New stardata with updated properties
+    """
```

Comment:
I am okay with ditching decamutil and going to WCS coordinates. The actual reference wavefront values are reported in pixel coordinates as well as focal_x and focal_y. This means they can be converted into (u,v).

In the Wavefront model, we are concerned with terms that modify zernike coefficients based on location in the focal plane. So then at given location (focal_x, focal_y) the zernike coefficient a_i[focal_x, focal_y] = a_i[value from knn interpolant] + delta_i + theta^x_i * focal_y + theta^y_i * focal_x, where the theta and delta terms are what we fit for each exposure. Based on what you are saying, we should be able to represent the focal_x and focal_y as some sort of transform on u,v. Then, if we wanted to compare our theta parameters with my old wavefront code and aaron's active optics system, we just have to do a unit conversion.